### PR TITLE
Pipeline State Objects

### PR DESCRIPTION
**Pull request by kvark**

### CODE REVIEW

File: _posts/2016-01-22-pso.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+How do we define it? One observation is that all of these 3 structs just carry some information for each PSO component: we could use the same field names and just store different data. Another observation is that all 3 are connected and can be generated from a single definition. E.g. for a constant buffer, the `Init` struct would have it's name (and potentially the layout of the elements), the `Meta` struct would have the constant buffer index to bind to, and the `Data` struct would just contain a `Buffer<T>`. We got a nice [table](https://github.com/gfx-rs/gfx/pull/828#issuecomment-160241713) of the components here for the reference.
+
+### Gfx-rs Solution
+The user defines a PSO in one places with a macro:
```
Comment (cmr) [inline _posts/2016-01-22-pso.md]: s/places/place

File: _posts/2016-01-22-pso.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
+---
+layout: post
+title: Pipeline State Objects
+---
+
+## Introduction
+
+PSO appears to be an integral part of the next-gen graphics APIs: DX12 [has them](https://msdn.microsoft.com/en-us/library/windows/desktop/dn899196%28v=vs.85%29.aspx), and [so does](https://developer.apple.com/library/ios/documentation/Metal/Reference/MTLRenderPipelineState_Ref/index.html) Metal. At Gfx-rs, we couldn't ignore this concept, and we went much further than simply adding support for it. We ditched our old abstractions (based around `Batch` concept) and put PSO at the center of the interface. This move has cleaned up Gfx-rs and brought it closer to the bare API level. The main challenge was making the PSO management type-safe and convenient without any run-time overhead.
```
Comment (cmr) [inline _posts/2016-01-22-pso.md]: I'd expand the first reference from PSO to "Pipeline state objects (PSO) appear to be..."

File: _posts/2016-01-22-pso.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+struct Init<'a> {
+    vertex: (),
+    const_locals: &'a str, // this is the shader-visible name
+    tex_diffuse: &'a str,
+    buf_noise: &'a str,,
+    pixel_color: &'a str,
+    depth: gfx::state::Depth,
+}
+{% endhighlight %}
+Since the `Init` one will need to be passed for PSO construction by the user, it's clear that our solution doesn't require (or even allow) the user to specify any redundant information, or miss anything. Let's say the user wants to enable blending for the `pixel_color`. All that needs to be done is changing the meta type from `RenderTarget<>` into `BlendTarget<>`. This would not change the corresponding `Data` component (which will still be `RenderTargetView<>`), but it would add a mandatory `gfx::state::Blend` to the `Init` struct field.
```
Comment (cmr) [inline _posts/2016-01-22-pso.md]: I'm not in general a fan of `Type<>`, and prefer `Type<T>`.

File: _posts/2016-01-22-pso.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+{% highlight Rust %}
+    let data = pipe::Data {
+        vertex: .., const_locals: ..,
+        tex_diffuse: .., buf_noise: ..,
+        pixel_color: .., depth: ..,
+    };
+    let slice = gfx::mesh::Slice {...};
+    encoder.draw(&slice, &pso, &data);
+{% endhighlight %}
+This is rather minimalistic, but, more importantly, there is next to no opportunity to shoot yourself in the foot! All the inputs of the PSO are guaranteed at compile time (if the type system is not abused). Performance-wise the solution is also perfect - all the mapping from the user data and the PSO inputs is already ready for us (contained in the `Meta`, which is constructed during PSO creation), so the actual binding doesn't involve any lookups.
```
Comment (cmr) [inline _posts/2016-01-22-pso.md]: What sort of abuse do you have in mind, for "if the type system is not abused"? It seems like an unnecessary comment

**Comment by cmr** [comment]
Looks good to me.

**Comment by kvark** [comment]
How about now?

**Comment by cmr** [comment]
@homu r+

**Comment by cmr** [comment]
Yeah, didn't expect to have homu here, with nothing to test :P

**Pull request merged**